### PR TITLE
fix putdata memory leak

### PR DESCRIPTION
**Pull request by benoit-pierre**
Fix for #1187

### CODE REVIEW

File: Tests/check_putdata_leaks.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env python
+
+from helper import unittest, PillowTestCase
+import sys
+from PIL import Image
+
+mem_limit = 8 # max increase in MB
+iterations = 50000
+test_file = "Tests/images/hopper.png"
```
Comment (homm) [inline Tests/check_putdata_leaks.py]: Doesn't used

File: _imaging.c [hunk 2]
```diff
@@ -1242,6 +1242,8 @@ _putdata(ImagingObject* self, PyObject* args)
         return NULL;
     }
 
+    seq = NULL;
```
Comment (homm) [inline _imaging.c]: Move this to declaration?

File: Tests/check_putdata_leaks.py (hunk 1 above)
Comment (benoit-pierre) [inline Tests/check_putdata_leaks.py]: Right, I changed the test after. Will amend.

**Comment by gareth-rees** [comment]
This change only fixes the leak in `putdata` — there's a similar leak in `getlist` and you might as well fix that at the same time.

File: _imaging.c (hunk 2 above)
Comment (benoit-pierre) [inline _imaging.c]: If you prefer, sure.

**Comment by benoit-pierre** [comment]
@gareth-rees : thanks, will amend to check for this too.

**Comment by benoit-pierre** [comment]
OK, for some reason the getlist test does not fail with Python 2. Any idea why?

I've converted the check to a proper test. Tell me if you prefer to keep it out of the testsuite for fear of false positives (the Travis build passed on my branch, hopefully it will pass for the pull request too).

**Comment by gareth-rees** [comment]
In Python 2, `range(256)` returns a list (not a range object, as in Python 3):
```
Python 2.7.9 (default, Dec 13 2014, 15:13:49) 
>>> type(range(256))
<type 'list'>
```
and so `PySequence_Fast` just increments its reference count and returns it:
```
    if (PyList_CheckExact(v) || PyTuple_CheckExact(v)) {
        Py_INCREF(v);
        return v;
    }
```
So the list is leaked, but there's only one list, and the leak is bounded in size. The test needs an argument that will make `PySequence_Fast` allocate a new list. `xrange` would do, except that it's not portable to Python 3, so you need something like:
```
# Need a sequence object that causes PySequence_Fast to allocate a list each time it's called.
try:
    _range = xrange # Python 2
except NameError:
    _range = range # Python 3
```
Alternatively, something like `bytearray(range(256))` would work, but it would need a comment to explain why you had to resort to this trickery!

**Comment by benoit-pierre** [comment]
Okay, thanks for the explanation!

File: Tests/test_imaging_leaks.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+min_iterations = 100
+max_iterations = 1000
+
+@unittest.skipIf(sys.platform.startswith('win32'), "requires Unix or MacOS")
+class TestWebPLeaks(PillowTestCase):
+
+    def _get_mem_usage(self):
+        from resource import getpagesize, getrusage, RUSAGE_SELF
+        mem = getrusage(RUSAGE_SELF).ru_maxrss
+        return mem * getpagesize() / 1024 / 1024
```
Comment (homm) [inline Tests/test_imaging_leaks.py]: `from __future__ import division` needs for this

File: Tests/test_imaging_leaks.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    def _test_leak(self, min_iterations, max_iterations, fn, *args, **kwargs):
+        start_mem = self._get_mem_usage()
```
Comment (homm) [inline Tests/test_imaging_leaks.py]: There is no need in `start_mem` anymore. `mem < mem_limit` constraint is enough.

**Comment by benoit-pierre** [comment]
I changed it to:

```python
self._test_leak(min_iterations, max_iterations,
                # Pass a new list at each iteration.
                lambda: im.point(range(256)))
```

File: Tests/test_imaging_leaks.py (hunk 3 above)
Comment (benoit-pierre) [inline Tests/test_imaging_leaks.py]: OK, note that this code is used elsewhere (in Tests/check_webp_leaks.py).

Comment (homm) [inline Tests/test_imaging_leaks.py]: Could you also add future division there?

**Comment by hugovk** [comment]
In general we've been avoiding putting DOS-type tests in the main test suite (i.e. naming them check_* instead of test_* -- test_* are picked up by our nose command), not so much that they may fail on Travis CI, but they cause issues on a resource-constrained environment.

**Comment by benoit-pierre** [comment]
@hugovk : I will rename to check_imaging_leaks.py.

File: Tests/test_imaging_leaks.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+from __future__ import division
+from helper import unittest, PillowTestCase
+import sys
+from PIL import Image, ImageFilter
+
+min_iterations = 100
+max_iterations = 10000
+
+@unittest.skipIf(sys.platform.startswith('win32'), "requires Unix or MacOS")
+class TestWebPLeaks(PillowTestCase):
```
Comment (hugovk) [inline Tests/test_imaging_leaks.py]: -> `class TestImagingLeaks(PillowTestCase):`

**Comment by hugovk** [comment]
If you want to get it passing on the CI first, you can rename it afterwards if you want.

File: Tests/test_imaging_leaks.py (hunk 5 above)
Comment (benoit-pierre) [inline Tests/test_imaging_leaks.py]: Yep, already fixed.

**Comment by VOTProductions** [comment]
Great work guys. :)

**Comment by hugovk** [comment]
Thanks!

**Pull request merged**